TCD-2905: ptrace (#2269)
saucelabs/sauce-docs: docs/error-reporting/advanced/ptrace.md

```lua
-- Example global variable iteration.
pmodule.log(pmodule.log_level.warning, "Global variables:")
for var, object, cu in bt:variables(), {object = "crash", cu = "invalid_write.c"} do
    pmodule.log(
        pmodule.log_level.warning,
        "name: %s, value: %s, object: %s, cu: %s",
        var:name(),
        tostring(var:value()),
        object,
        cu
    )
end

-- Example TLS variable iteration.
for thr in bt do
    pmodule.log(pmodule.log_level.warning, "TLS variables:")

    for var, object, cu in thr:variables(), {cu = "hang"} do
        pmodule.log(
            pmodule.log_level.warning,
            "name: %s, value: %s, object: %s, \z
                cu: %s",
            var:name(),
            tostring(var:value()),
            object,
            cu
        )
    end
end
```

These iterations use the same generic `for` interface mentioned earlier for threads, frames, and other entities. The filter (invariant state) here is a table instead of a match object. The supported fields are:

- `object`: The desired object file (substring match).
- `cu`: The desired compilation unit (substring match).
- `name`: The variable's name (exact match).

> All of these fields (and the filter table itself) are optional. Each iteration returns a variable object, the object file name, and the compilation unit name.

Author: ```suggestion
All of these fields (and the filter table itself) are optional. Each iteration returns a variable object, the object filename, and the compilation unit name.
```

    print(string.format("addr: %x, len: %d, size: %d", addr, n, size))

    for i = 0, n - 1 do
        local na = addr + i * size

        print(string.format("--- iter %d (%x)---", i, na))

        elem = pmodule.deref(cached_ref_var, addr + i * size)

        print(string.format("a: %d", pmodule.variable_from_bt(elem.a):value()))
        print(string.format("b: %f", pmodule.variable_from_bt(elem.b):value()))

        -- Note: when accessing member fields, pointers are
        -- automatically dereferenced until we reach the underlying
        -- struct.
        print(string.format("nppp.c: %d", pmodule.variable_from_bt(elem.nppp.c):value()))
    end
end

local function pm_load()
    local m = pmodule.match()

    m:add_object("crash")
    m:add_file("crash.c")
    m:add_variable_base_type("array_struct", pmodule.match_type.substr)
    m:add_variable_ptrace_type(pmodule.variable_type.tuple)
    m:add_variable_name("papa", pmodule.match_type.substr)

    pmodule.register(pmodule.event.variable, variable_cb, m)
    pmodule.register(pmodule.event.attach, attach_cb)
end

pmodule.define {
    id = "array_iter",
    load = pm_load
}
```

> There are a few CLI options to keep in mind when using pmodules:

Author: ```suggestion
A few CLI options exist to keep in mind when using pmodules:
```
